br/pkg/lightning/restore/restore.go

```diff
@@ -277,37 +279,58 @@ type LightningStatus struct {
 	TotalFileSize    atomic.Int64
 }
 
+// ControllerParam contains many parameters for creating a Controller.
+type ControllerParam struct {
+	// databases that dumper created
+	DBMetas []*mydump.MDDatabaseMeta
+	// a pointer to status to report it to caller
+	Status *LightningStatus
+	// storage interface to read the dump data
+	ExtStorage storage.ExternalStorage
+	// if ExtStorage is created by lightning. In some cases where lightning is a library, the framework may pass an ExtStorage
+	OwnExtStorage bool
+	// used by lightning server mode to pause tasks
+	Pauser *common.Pauser
+	// lightning via SQL will implement its glue, to let lightning use host TiDB's environment
+	Glue glue.Glue
+	// when not OwnExtStorage, checkpoint can also be saved in framework-created ExtStorage by setting this field
+	CpNameInExtStorage string
```

Comment:
Lightning assumes all files in external storage are source data. I think it's not a good way to put checkpoint file in it, although we can use file route to ignore specific file. I suggest using a separate external storage to store checkpoint. If you really want to use the same external storage, you can pass the same object.

Reply:
This is already the behaviour of lightning in DM https://github.com/pingcap/tiflow/pull/3813

Reply:
How does DM works without this PR before?

Reply:
you can search https://github.com/pingcap/tiflow/pull/3813/files

https://github.com/pingcap/tiflow/pull/3813/files#diff-dc63bb506f94f15965a3d2fed1dcf08ced7e068fb8b07488f90895dfd40e0e38R196

```golang
		cfg.Checkpoint.Driver = lcfg.CheckpointDriverFile
		cpPath := filepath.Join(l.cfg.LoaderConfig.Dir, lightningCheckpointFileName)
		cfg.Checkpoint.DSN = cpPath
```

Reply:
It seems it just set checkpoint path inside the source dir, but it doesn't limit it set to other path.

Using the external storage for source data and checkpoint has potential risks that the checkpoint file may be treated as a source file to import. So I think we shouldn't put checkpoint in the same external storage. But if DM still want to put them together, two external storage can be created. e.g source dir: `file:///dump`  checkpoint: `file:///dump/checkpoint`.

Reply:
cc @glorv

Reply:
> but it doesn't limit it setting to other path

by the snippet above, the DSN is limited to a hardcoded checkpoint file under dump file directory. I remember the main purpose is to let checkpoint and dump data exist and lose at same time.

> source dir: file:///dump checkpoint: file:///dump/checkpoint

I remember lightning will scan the dump data recursively

Reply:
>  the main purpose is to let checkpoint and dump data exist and lose at same time.

Lightning doesn't require  checkpoint and dump data exist at the same time. If lightning has already restored data, the dump data can be removed. For Lightning binary, the source dir may be read only and we can't write anything in it. Although this PR is for library usage currently, I would like to lightning binary and library use the same implementation later.

> I remember lightning will scan the dump data recursively
Yes. So I'm not in favor of putting them in the same directory.

Reply:
I think most of concerns are about DM/dataflow engine should be more careful when storing file checkpoint in dump folder, not related to lightning. For example,

> the checkpoint file may be treated as a source file to import.

In DM or dataflow engine's case, dumpling is used to create dump files, so we may limit the dump file matching pattern with this apriori knowledge to avoid bugs.

But if we actually store them at two place, (which requires two different arguments for two storage parameters), lightning has to handle the case that only one of them exists. And risk is higher since it's rely on two storages are available rather than one.

Reply:
> lightning has to handle the case that only one of them exists. 

Maybe we can let the two storages required and disallow one of them is nil. Or create `checkpointsDB` out of `NewRestoreController` and pass `checkpointsDB` to it.

Reply:
> > lightning has to handle the case that only one of them exists.
> 
> Maybe we can let the two storages required and disallow one of them is nil. Or create `checkpointsDB` out of `NewRestoreController` and pass `checkpointsDB` to it.

Oh I mean files on one of two storages are cleaned. Can lightning report error when only checkpoint exists? And lost checkpoint will cause some data restarted, which will not happen if we store them together.

Reply:
Maybe not. I think these are two concerns.
1. For Lightning, I don't hope checkpoint can only store in dump dir's external storage. 
2. For DM, If you want to avoid "one of them loss", you can pass the same external storage object to both dump dir and checkpoint path.